llext: cache section headers

pillo79 opened the pull request:
The `llext_load()` and `llext_link()` functions need to access section header data from the ELF file quite often, but this is currently done by seeking and reading each section header whenever it becomes necessary (even lots of times for the same section, in the case of symbols). 

This PR reworks the current code by loading the whole section table in memory and accessing the array elements directly in the code. The memory for the section headers is freed at the end of the `llext_load()` function. If `llext_peek()` is supported, no allocation is performed and the data is fetched directly from the ELF file buffer.

* The first commit is cosmetic, and restores some debug messages which were lost in the split-file refactor.
* The second commit in the series refactors the code freeing memory at the end of `llext_load()` so that it's clearer and more explicit. As an improvement, the symbol table is kept in memory in debug builds to ease debugging.
* The third commit adds the code that loads this section header array and frees it (or uses it directly from the ELF file buffer if available).
* The last commit is a pure refactor that introduces no logic changes, but removes all code reading section headers and replaces it with the array indexing.

pillo79 committed:
llext: llext_load: document memory management policy

The do_llext_load function is responsible for loading an extension from a
file, and for this purpose it calls a number of functions that a) allocate
memory, and b) can fail. This creates the opportunity for memory leaks if
the error paths are not handled correctly.

This commit adds a comment at the beginning of the function to document
the memory management policy that has to be followed in this file:
cleanup is not performed in the error paths, and all memory is freed at
the end of the do_llext_load() function, both in the case of error and of
successful loading.

As an improvement, the symbol table is not freed if the LLEXT log level
is set to debug, so that it can be used, for example, to inspect the
symbols of the loaded extension.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: improve debug messages

The recent llext_map_sections() rework changed the way debug messages
are output so that the names of most skipped sections are not printed
at all. This makes debugging harder since the section names are useful
to identify the contents at a glance.

Also print a few additional fields from the section header, and use 0x
prefixes for hex numbers.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: refactor: use cached section headers

The section headers are now available in the loader structure, so we can
use those directly instead of reading them from the ELF file every time.

This commit contains no logic changes; it removes all copies of the
header loading code and replaces them with direct access to the cached
section headers.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

mathieuchopstm commented on subsys/llext/llext_load.c:
```diff
@@ -539,6 +526,37 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext,
+		llext_free(ext->exp_tab.syms);
+		ext->exp_tab.sym_cnt = 0;
+		ext->exp_tab.syms = NULL;
+	}
```
Nit: I would regroup symbol table freeing under a single `if (ret != 0 || !IS_ENABLED(CONFIG_LLEXT_LOG_LEVEL_DBG))`.
In current state, I find it difficult to see the difference between this block and the one above (this one also calls `llext_free_section`).

pillo79 replied:
I know it's difficult to catch `sym_tab` (ELF symbols, freed unless in debug mode) vs `exp_tab` (llext public symbols, freed in case of error). I will definitely improve comments if I need to do a second iteration: the debug one should have been outside the `if` :man_facepalming:

mathieuchopstm reviewed (approved):
LGTM, with a minor cosmetic complaint.

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -542,6 +542,32 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext,
+	ldr->sect_map = NULL;
+
+	if (!IS_ENABLED(CONFIG_LLEXT_LOG_LEVEL_DBG)) {
+		/* keep the computed symbol tables for debugging */
```
what if this is called in an error case? Maybe swap around as
```
	if (ret != 0) {
		/* free sections and exported symbols */
		...
	} else if (IS_ENABLED(CONFIG_LLEXT_LOG_LEVEL_DBG)) {
		return;
	}
	/* free the symbol table */
	...
```

pillo79 replied:
Indeed, thanks for spotting this! I will refactor this appropriately.

mathieuchopstm replied:
In error case, `ext->exp_tab.syms` is "free'ed twice", but is already NULL on second `free` so nothing bad happens (`free(NULL)` is no-op in Zephyr).
I agree with your comment nonetheless (and made the same one a few lines below :smile:)

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -97,12 +108,36 @@ static int llext_load_elf_data(struct llext_loader *ldr, struct llext *ext)
+
+		ret = llext_seek(ldr, ldr->hdr.e_shoff);
+		if (ret != 0) {
+			LOG_ERR("Failed to seek for section headers");
```
this immediately looks like a memory leak. I think it's better to free all partially allocated resources *inside* the function that allocated them but failed to complete successfully. This just feels more robust to me: either the function succeeds and everything that it had to allocate is allocated, or it failed and returned the system to the state before the call. At the very least I'd add comments, saying that the caller will free...

teburd replied:
I tend to agree but in this case we *do* cleanup, just one level up in the call stack no?

pillo79 replied:
That also applies to almost every other error path in  @this file. :slightly_smiling_face: It was purposefully done to avoid littering the code with `free`s and/or `goto`s, which I find a lot more complicated to verify. My rationale was:
- all pointers are NULL at the beginning of `llext_load` (zeroed out `struct llext`); 
- error out liberally in `do_llext_load` and depending functions, whatever the stage;
- free stuff appropriately once, in `llext_load_cleanup`, according to the result.

Would adding a very big comment describing this be OK?

lyakh reviewed (dismissed):
this *looks* good but this time I **haven't tested** it with SOF

teburd commented on include/zephyr/llext/loader.h:
```diff
@@ -71,6 +71,8 @@ struct llext_loader {
 	/** @cond ignore */
 	elf_ehdr_t hdr;
 	elf_shdr_t sects[LLEXT_MEM_COUNT];
```
Is this still needed given we allocate/reference the section headers below? Should we use pointers here instead?

pillo79 replied:
Those are still useful since they host the "merged" sections, and must be writable. I think it _could_ be possible to use pointers in specific cases, but you would be modifying the original values which are used by the linking later. 
I would leave this for a further optimization.

teburd commented on subsys/llext/llext_load.c:
```diff
@@ -112,56 +139,42 @@ static int llext_load_elf_data(struct llext_loader *ldr, struct llext *ext)
 			if (ldr->hdr.e_shstrndx == i) {
 				LOG_DBG("shstrtab at %d", i);
-				ldr->sects[LLEXT_MEM_SHSTRTAB] = shdr;
+				ldr->sects[LLEXT_MEM_SHSTRTAB] = *shdr;
```
Do we need to copy the section headers here or can we directly reference them?

pillo79 replied:
See other comment - yes, we do need the copies for the other non-string sections.

teburd commented on subsys/llext/llext_load.c:
```diff
@@ -625,20 +643,14 @@ int do_llext_load(struct llext_loader *ldr, struct llext *ext,
 
 out:
-	llext_free(ldr->sect_map);
+	llext_load_cleanup(ldr, ext, ret);
```
I'm not too fond of passing the ret into a function that does branching on it. I'd rather see the branch at this level myself. Something like...

```c 
llext_free_loader(ldr, ext);
if (ret != 0) {
   LOG_ERR("Failed to load extension...");
   llext_free_extension(ext);
} else {
  LOG_DBG("Loaded extension ...");
}
return ret;
```

pillo79 replied:
I see your point. Will merge the code in llext_load_cleanup after the label.

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -136,15 +136,17 @@ static int llext_find_tables(struct llext_loader *ldr)
-		LOG_DBG("section %d at %zx: name %d, type %d, flags %zx, "
-			"ofs %zx, addr %zx, size %zd",
-			i, pos,
+		LOG_DBG("section %d at 0x%zx: name %d, type %d, flags 0x%zx, "
```
Also done on purpose, to clearly differentiate integers and hex numbers. That message was the source for quite a few headaches for my brain! :grin:

pillo79 reviewed (commented):
Thanks for the review, will improve the comments and fix the issue.

pillo79 reviewed (commented):
Thanks @teburd, cross-posted :slightly_smiling_face: . Will refactor!

mathieuchopstm commented:
Unrelated to this PR, but this caught my attention while reviewing:
https://github.com/zephyrproject-rtos/zephyr/blob/dfb240f37ced92b5f487179dba6c93bc8f6c04e6/subsys/llext/llext.c#L160-L173

`unsigned int ret` is used as returned value for `int` function... is that OK?

pillo79 committed:
llext: read all section headers at once

This change reads all section headers at once, instead of reading them
one by one. This is more efficient and allows to further simplify the
code downstream.

The section headers are directly accessed from the file buffer if the
llext_peek() function is supported by the loader. Otherwise, they are
read into a buffer allocated on the heap and used only during the
llext_load() function.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
v2:
- added a lot more comments to the file header and in the code to explain the error management policy
- moved the cleanup code directly inside `llext_load()`
- fixed the memory leak for the symbol table on error spotted by @lyakh
